### PR TITLE
Dynamic ruleset threat scaling and various fixes

### LINKED ISSUE #46218: I became a headrev as a cyborg on a dynamic round
I was playing as a cyborg (I had been borged, I didn't start the round as a cyborg) and I suddenly got a blue R on my mob, and when I flashed people, they became revs. However, I had no text informing me "you are the head revolutionary" or anything similar. It really confused the crew!
Issue reported from  Round ID: 117284 (/tg/Station Terry [EU] [100% LAG FREE] [New-Dedi])

Reporting client version: 512.1477

### LINKED ISSUE #47234: revs gamemode not clearing hostile shuttle environment flag on dynamic
When all the headrevs are dead it's not properly clearing the revolution hostile environment mode.

Terry, round id 121804

### LINKED ISSUE #47283: Game can choose players that are not on station z-level as revheads
Issue reported from  Round ID:  121914 (/tg/Station Terry [EU] [100% LAG FREE] [New-Dedi])

Both revheads were miners on lavaland. Round ended instantly. Game should ignore people that are not on station when rolling revheads.

Reporting client version: 512.1471

### PR DESCRIPTION
**Pull request by kriskog**
Do not merge this without a testmerge, prefereably on Terry.

Dynamic changes during antag freeze approved by oranges.

## About The Pull Request
Rewritten the pop scaling and added threat scaling to the roundstart solo antag modes. Pop scaling is now a fixed number per requirement, which makes it a bit easier to adjust directly by non-coders in the config. The roundstart traitor, blood bro and changeling rulesets now all have a low base cost, and will instead attempt to scale up their number of antags spawned based on threat left and amount of other roundstart modes. This should for the most part leave less room in available threat to pick the round-ender modes as 2nd or 3rd mode when one of the above rulesets are picked first, and instead put more threat focus into those modes. This also means we can get more varied combinations of antags, instead of modes like traitor, blood bros and changelings always spawning the max pop-supported amount once they are picked.

Overall this should hopefully bring the absolute chaos factor down a little and spawn somewhat fewer, but more initial ruleset focused antags. 

I probably need to make some adjustments for the scaling calculation, but I want some actual live data to see how this performs. Further adjustments might also be made through the configs, I will be working with the headmins on that as this is tested.

Also weeded out a few bugs with forced rulesets, and made the config actually work out of the box.

Todo:
 - [x] Make changes to revs ruleset to fit better with dynamic and additional antags.
 - [x] Do blood bros pop scaling right - currently only counts the team as one antag for future scaling purposes
 - [x] Add better remaining rule validation upon picking a rule.
 - [x] Make sure the math makes sense. 
 - [x] Remove wizard events from dynamic?

## Why It's Good For The Game
Chaos is fun and good, but too much chaos all the time bad. Reigns dynamic in a little.

Fixes #47234
Closes #46887
Fixes #46218
Fixes #47283

## Changelog
:cl: Skoglol, TheChosenEvilOne
add: Dynamic now has a more threat based scaling of the individual solo antag modes, instead of only scaling amount of modes to threat. This should make for a somewhat less chaotic jumble of rulesets, with later picked ones supporting the initial one instead of overshadowing every time. Old style chaos still possible though.
admin: Dynamic: Configs now work. 
add: Dynamic: Second and third rulesets rolling now has a threat-based probability of rolling instead of always doing so.
fix: Dynamic: Rulesets should now fire more consistently.
fix: Dynamic: Round-ender rulesets now (properly) cant stack below 90 threat. Several bugs fixed.
fix: Dynamic: Fixed bugs related to forcing normal rulesets.
tweak: Dynamic: Revolutions will not end the round on a station win.
fix: Dynamic: Revolution should now end properly.
fix: Dynamic: Revolution will no longer make ineligible players revheads. Borgs, get back in line.
balance: Dynamic: Wizard summon events/guns/magic disabled.
balance: Dynamic: Blood cult ruleset is now 35 cost by default.
/:cl:

### CODE REVIEW

**Comment by Iatots** [comment]
Am I reading this correctly when I say that there can now be a max of 4 wizards roundstart?

**Comment by kriskog** [comment]
You are not, there is a skip in the file preview and you are reading off from blood cult.

**Comment by Ty-the-Smonk** [comment]
Wiz events should end if the wizard dies

**Comment by Robustin** [comment]
Would you mind breaking down exactly how this would affect odds on a typical ~70 person round?

**Comment by kriskog** [comment]
I will do that, probably tomorrow. TCEO had some 2nd and 3rd ruleset probabilities in the pipeline, I need to decide if I wish to keep those (which changes stuff up some more, need to redo the math) or scrap them and leave it all to scaling.

**Comment by kriskog** [comment]
Allright, here are some numbers and lots of text to illustrate how this will work. All these values are subject to change during testing.

First off, ruleset probability added by TCEO based on threat. Index every 10 threat. The probs are fairly high, so this wont do THAT much.
```
var/list/second_rule_prob = list(0,0,60,80,80,80,100,100,100,100)
var/list/third_rule_prob = list(0,0,0,0,60,60,80,90,100,100)
```
This means that the second ruleset is less likely at lower threat, and the third one even less so. The amount of rulesets does affect the probability of scaling the solo antag rulesets, as you will see later.

First off, here is the math for scaling.
base_prob + remaining_threat_level - (scaled_times * scaling_cost) - (extra_rulesets * EXTRA_RULESET_PENALTY) - (pop_to_antags * POP_SCALING_PENALTY)
Or with numbers where applicable yet:
60 + remaining_threat_level + (scaled_times * scaling_cost) - (extra_rulesets * 15) - (pop_to_antags * 50)

Lets assume a 50 pop round rolling 50 threat, which should be a fairly average round. This would qualify for both the second and third rulesets based on pop, and for threat there is an 80% chance the second ruleset is picked and a 60% chance of the third. Let's assume the third fail, so we got two rulesets.

If the first ruleset picked is a bigger team antag mode, everything is as before with no scaling. In the example of cult, that's 30 threat spent on cult. The second ruleset then has 20 threat to spend, which leaves traitor (10), blood bro (15) or changeling (15). Traitor is low enough that it would attempt to scale if picked, while the others cant afford it, so lets look at how that works out.

The numbers for traitor scaling would then be:
60 + 10 (remaining threat) - 0 (scaled 0 times before) - 0 (no more rulesets after) - 7 (already got 7 antags, math equalizes at 50 pop) = 63 percent chance of scaling up one more step.

So in a bit over half the cases, we would have a cult + 6 traitors. Current system would have been forced to grab traitors due to cost for second ruleset, and would have spawned 6 traitors as well with 10 threat left over. Do note that while the scaling systems traitor population cap has already been reached at 3 per step, the old traitor ruleset would keep adjusting up based on pop (ie 12 traitors at 80 pop).
This isn't where scaling REALLY kicks in though, since cult is the "main" mode this round and spent most of our threat.

Same parameters, 50 pop with 50 threat. Two rulesets available, traitor rolls first and attempts to scale up since there is spare threat. At 50 pop, 3 traitors will spawn per step.

60 + 40 - 0 - 15 - 3 = 82

So the first scale up is likely. -10 threat, we are now at 30 left.

60 + 30 - 10 - 15 - 6 = 59

Second scale up is less likely, but will happen more often than not. If it does, all the round enders are now too expensive to be run and we got a more solo antaggy round. Lets assume it scaled the second time.

60 + 20 - 20 - 15 - 9 = 36

So the third scale up is unlikely. The rulesets can not scale more than three times. Lets assume it didn't.

We got 20 threat left, which leaves us with brothers or changelings both at 15 cost. If brothers spawn, we get one team of brothers on top of 9 traitors. If changelings spawn, we get 3 changelings.

With the current system, the same rolls would equate to either: 
6 traitors (10), and either a 30-40 cost round ender ruleset (cult, nukies, revs), wizard, bloodbros or changelings. In the case of either blood bros or changelings, 2 teams or 6 changelings.

So the amount of antags from the "solo" focused rulesets would be pretty similar, except we spent a little bit more threat. The big change is in how the round enders were forced out by the scaling, and we got a round mostly consisting of traitors for once.


Again, pending testing and some real data on this, the numbers may change. These numbers sound reasonable to me at around this pop, but the lower pop rounds are where they might get iffy. I got a hard cap on 50% antags before scaling will be ignored, and that in conjunction with the ruleset probability and general pop limits might be enough. We will see if/when we get this testmerged.

The goal is also to keep these numbers adjustable via configs in the future, in case someone wishes to take the time to experiment. The easiest way to do this would be adjusting the rulesets base prob and scaling cost.

tl;dr scaling good, round enders bad (when all of the time)

**Comment by TheChosenEvilOne** [comment]
Just a quick thing you should do is to move the persistent rule check from picking_roundstart_rule to post_setup to prevent the runtime that happens with revolution when it tries to process the ruleset when the revolution team does not exist.

**Comment by kriskog** [comment]
Not touching this anymore for now pending review and testmerge.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 1]
```diff
@@ -374,12 +373,32 @@
...
+				if(check_eligible(newrev.mind))
+					break
+			assigned += newrev.mind
+			newrev.mind.restricted_roles = restricted_roles
+			newrev.mind.special_role = antag_flag
+			var/datum/antagonist/rev/head/new_head = new antag_datum()
+			new_head.give_flash = TRUE
+			new_head.give_hud = TRUE
+			new_head.remove_clumsy = TRUE
+			newrev.mind.add_antag_datum(new_head,revolution)
```
Comment (TheChosenEvilOne) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]: This will not work as candidates is already an empty list and if it wasn't it would be probably just list of nulls.

Comment (TheChosenEvilOne) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]: And anyways just having one head is enough as then the system can "promote" people.

Comment (kriskog) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]: Ah, hm yes you are right. Fuck it then, thanks.

**Comment by Iatots** [comment]
Revolution succeeding could still end the round prematurely for other antags. How about if revolution wins, the station goes into egalitarian mode and the shuttle is called? Might prevent edge cases like #47234

**Comment by kriskog** [comment]
I have considered something like that, but it requires some more work than I want to add onto this PR. It is already bigger than I intended.
The edge case you linked should already be fixed in this PR though.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm (hunk 1 above)
Comment (TheChosenEvilOne) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm]: If nothing else if there are not enough valid revheads it should refund the ruleset and then kill the processing.

File: code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm [hunk 2]
```diff
@@ -89,27 +90,82 @@
 	requirements = list(101,101,70,40,30,20,20,20,20,20)
 	high_population_requirement = 50
 	flags = HIGHLANDER_RULESET
-	var/required_heads = 3
+	var/required_heads_of_staff = 3
+	// 1 for revolution win and 2 for heads of staff win.
+	var/finished = 0
```
Comment (AnturK) [inline code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm]: defines

File: code/modules/antagonists/wizard/equipment/spellbook.dm [hunk 3]
```diff
@@ -492,6 +492,8 @@
 /datum/spellbook_entry/summon/guns/IsAvailible()
 	if(!SSticker.mode) // In case spellbook is placed on map
 		return FALSE
+	if(SSticker.mode == /datum/game_mode/dynamic) // Disable events on dynamic
```
Comment (AnturK) [inline code/modules/antagonists/wizard/equipment/spellbook.dm]: You want istype here and below

File: code/modules/mob/living/carbon/human/death.dm [hunk 4]
```diff
@@ -45,11 +45,16 @@
...
 
 
 /mob/living/carbon/proc/Drain()
 	become_husk(CHANGELING_DRAIN)
 	ADD_TRAIT(src, TRAIT_BADDNA, CHANGELING_DRAIN)
 	blood_volume = 0
-	return 1
+	return TRUE
+
+/mob/living/carbon/proc/makeUncloneAble()
```
Comment (AnturK) [inline code/modules/mob/living/carbon/human/death.dm]: Sane capitalization please.

File: config/game_options.txt [hunk 5]
```diff
@@ -201,13 +201,6 @@ MIDROUND_ANTAG WIZARD
 ## The amount of time it takes for the emergency shuttle to be called, from round start.
 SHUTTLE_REFUEL_DELAY 12000
 
-## Variables calculate how number of antagonists will scale to population.
-## Used as (Antagonists = Population / Coeff)
-## Set to 0 to disable scaling and use default numbers instead.
-TRAITOR_SCALING_COEFF 6
```
Comment (AnturK) [inline config/game_options.txt]: Why remove these ?

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/death.dm (hunk 4 above)
Comment (kriskog) [reply code/modules/mob/living/carbon/human/death.dm]: Was gonna say something about following the file standard, but its mixed. Underscores time.

Comment (AnturK) [reply code/modules/mob/living/carbon/human/death.dm]: We don't really have a standard but random capital letter in the middle is just too jarring.

File: config/game_options.txt (hunk 5 above)
Comment (kriskog) [reply config/game_options.txt]: They are no longer in use. The rulesets now have fixed amounts of antags per pop incide instead, as well as individual scaling probability, antag cap and scaling cost configs in the dynamic.json file.

File: code/modules/mob/living/carbon/human/death.dm (hunk 4 above)
Comment (kriskog) [reply code/modules/mob/living/carbon/human/death.dm]: Uncloneable is one word isnt it, oof

File: config/game_options.txt (hunk 5 above)
Comment (AnturK) [reply config/game_options.txt]: These are not dynamic only though.

File: code/modules/mob/living/carbon/human/death.dm (hunk 4 above)
Comment (kriskog) [reply code/modules/mob/living/carbon/human/death.dm]: I usually just keep to whatever the above procs is done as, to keep consistent with similar procs. This was just a language failure on my part.

File: code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm (hunk 2 above)
Comment (kriskog) [reply code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm]: done

File: code/modules/antagonists/wizard/equipment/spellbook.dm (hunk 3 above)
Comment (kriskog) [reply code/modules/antagonists/wizard/equipment/spellbook.dm]: done

File: config/game_options.txt (hunk 5 above)
Comment (kriskog) [reply config/game_options.txt]: Oh god oh fuck. Reverted, thanks!
